### PR TITLE
✨ Add EKS bootstrap provider

### LINKED ISSUE #1779: EKS bootstrap provider
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

EKS support will require different userdata from that created by the Kubeadm provider. Because of this, we should create a new bootstrap provider.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

EKS bootstrap provider should be created in the CAPA repo at `bootstrap/eks` to match standards set by the CAPI repo.

I'll add some more implementation considerations to this issue shortly.

### PR DESCRIPTION
**Pull request by rudoi**
**What this PR does / why we need it**:

This PR introduces the EKS bootstrap provider. It includes a new CRD, `EKSConfig`.

Initial implementation of this provider makes the assumption that the [official EKS bootstrap script](https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh) is present on the AMI at `/etc/eks/bootstrap.sh`.

This assumption will not really be appropriate until #1778 is hooked up.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1779 

Turns out creating a draft PR doesn't work so well with the robots, so I had to reopen.

### CODE REVIEW

**Comment by NicoleYson** [comment]
I signed it 
😄 🤞

File: bootstrap/eks/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Build the manager binary
```
Comment (rudoi) [inline bootstrap/eks/Dockerfile:1]: this file will go away pending some Makefile/Dockerfile changes in the root

**Comment by rudoi** [comment]
cc @richardcase

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	if !scope.Cluster.Status.InfrastructureReady {
+		scope.Logger.Info("Cluster infrastructure is not ready, requeueing")
+		conditions.MarkFalse(scope.Config,
+			bootstrapv1.DataSecretAvailableCondition,
+			bootstrapv1.WaitingForClusterInfrastructureReason,
+			clusterv1.ConditionSeverityInfo, "")
+		return ctrl.Result{}, nil
+	}
+
+	if !scope.Cluster.Status.ControlPlaneInitialized {
```
Comment (richardcase) [inline bootstrap/eks/controllers/eksconfig_controller.go:162]: I need to double check we are setting the right status.

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *EKSConfigReconciler) SetupWithManager(mgr ctrl.Manager, option controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&bootstrapv1.EKSConfig{}).
+		WithOptions(option).
+		Complete(r)
```
Comment (richardcase) [inline bootstrap/eks/controllers/eksconfig_controller.go:193]: For the `if !scope.Cluster.Status. ` checks above will you be adding an additional watch here on `Cluster` so that reconcile is called on an update to it?

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		scope.Logger.Info("Cluster has not yet been initialized, requeueing")
+		conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretAvailableCondition, bootstrapv1.DataSecretGenerationFailedReason, clusterv1.ConditionSeverityWarning, "")
+		return ctrl.Result{}, nil
+	}
+
+	scope.Logger.Info("generating userdata", "cluster", scope.Cluster.GetName())
+
+	// generate userdata
+	userDataScript, err := userdata.NewNode(&userdata.NodeInput{
+		ClusterName:      scope.Cluster.GetName(),
```
Comment (detiber) [inline bootstrap/eks/controllers/eksconfig_controller.go:172]: I'm assuming that this relates to the underlying EKS Cluster name, if so, do we expect it to actually match the CAPI Cluster Name?

It might be good to come up with a convention here to not get into yet another situation where we could have conflicts on Cluster Name across namespaces....

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+		KubeletExtraArgs: scope.Config.Spec.KubeletExtraArgs,
+	})
+	if err != nil {
+		scope.Error(err, "Failed to create a worker join configuration")
+		return ctrl.Result{}, err
+	}
+
+	// store userdata as secret
+	if err := r.storeBootstrapData(ctx, scope, userDataScript); err != nil {
+		scope.Error(err, "Failed to store bootstrap data")
```
Comment (detiber) [inline bootstrap/eks/controllers/eksconfig_controller.go:182]: It would probably be good to update `DataSecretAvailableCondition` here with relevant information.

File: bootstrap/eks/hack/boilerplate.go.txt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/*
```
Comment (detiber) [inline bootstrap/eks/hack/boilerplate.go.txt:1]: Is this file needed, or can any scripts referencing it be updated to use the one in the root of the repo?

File: bootstrap/eks/internal/userdata/node.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	ClusterName      string
+	KubeletExtraArgs map[string]string
+}
+
+// NewNode returns the user data string to be used on a node instance.
+func NewNode(input *NodeInput) ([]byte, error) {
+	tm := template.New("Node")
+
+	if _, err := tm.Parse(argsTemplate); err != nil {
+		return nil, errors.Wrap(err, "failed to parse args template")
```
Comment (detiber) [inline bootstrap/eks/internal/userdata/node.go:43]: Does this actually need `errors.Wrap` or can we use the native go wrapping: `fmt.Errorf("failed to parse args template: %w", err)`?

File: test/helpers/envtest.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	Config *rest.Config
+
+	doneMgr chan struct{}
+}
+
+// NewTestEnvironment creates a new environment spinning up a local api-server.
+//
+// This function should be called only once for each package you're running tests within,
+// usually the environment is initialized in a suite_test.go file within a `BeforeSuite` ginkgo block.
+func NewTestEnvironment() *TestEnvironment {
```
Comment (detiber) [inline test/helpers/envtest.go:82]: Are we missing helpers like this in capi proper?

File: bootstrap/eks/main.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+		Scheme:             scheme,
+		MetricsBindAddress: metricsAddr,
+		LeaderElection:     enableLeaderElection,
+		LeaderElectionID:   "eks-bootstrap-manager-leader-election-capa",
+		LeaseDuration:      &leaderElectionLeaseDuration,
+		RenewDeadline:      &leaderElectionRenewDeadline,
+		RetryPeriod:        &leaderElectionRetryPeriod,
+		Namespace:          watchNamespace,
+		SyncPeriod:         &syncPeriod,
+		NewClient:          newClientFunc,
```
Comment (detiber) [inline bootstrap/eks/main.go:133]: This should be able to use `util.ManagerDelegatingClientFunc` from capi

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+		}
+		log.Error(err, "Failed to get config")
+		return ctrl.Result{}, err
+	}
+
+	// check owner references and look up owning Machine object
+	configOwner, err := bsutil.GetConfigOwner(ctx, r.Client, config)
+	if apierrors.IsNotFound(err) {
+		// no error here, requeue until we find an owner
+		return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:83]: Decided to go with watches for Machine and MachinePool (given the feature is enabled)

File: test/helpers/envtest.go (hunk 8 above)
Comment (rudoi) [reply test/helpers/envtest.go:82]: I honestly forget why we copied this to here. I was able to remove it and rely solely on CAPI helpers.

**Comment by rudoi** [comment]
ok @richardcase @detiber just pushed up a bunch of goodies lmao :D i think i addressed all existing comments.

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+				rerr = err
+			}
+		}
+	}()
+
+	return r.joinWorker(ctx, scope)
+}
+
+func (r *EKSConfigReconciler) joinWorker(ctx context.Context, scope *EKSConfigScope) (ctrl.Result, error) {
+
```
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:157]: ah yep, missed this comment, will address

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+		conditions.MarkFalse(scope.Config,
+			bootstrapv1.DataSecretAvailableCondition,
+			bootstrapv1.WaitingForClusterInfrastructureReason,
+			clusterv1.ConditionSeverityInfo, "")
+		return ctrl.Result{}, nil
+	}
+
+	if !scope.Cluster.Status.ControlPlaneInitialized {
+		scope.Logger.Info("Cluster has not yet been initialized, requeueing")
+		conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretAvailableCondition, bootstrapv1.DataSecretGenerationFailedReason, clusterv1.ConditionSeverityWarning, "")
```
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:184]: Yep agreed, will fix

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	scope := &EKSConfigScope{
+		Logger:      log,
+		Config:      config,
+		ConfigOwner: configOwner,
+		Cluster:     cluster,
+	}
```
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:126]: So, getting rid of the EKSConfigScope type entirely? I think I agree. We initially copied this to follow the pattern set forth in the Kubeadm provider but the logic in this controller is simpler so it may be a bit much.

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		log.Error(err, "Failed to get owner")
+		return ctrl.Result{}, err
+	}
+	if configOwner == nil {
+		// no error, requeue until we find an owner
+		return ctrl.Result{}, nil
+	}
```
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:97]: Added a watch on Machine/MachinePool to account for this.

File: bootstrap/eks/api/v1alpha3/eksconfig_webhook.go [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+func (r *EKSConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// (user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+// +kubebuilder:webhook:verbs=create;update,path=/validate-bootstrap-cluster-x-k8s-io-v1alpha3-eksconfig,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=bootstrap.cluster.x-k8s.io,resources=eksconfigs,versions=v1alpha3,name=validation.eksconfig.bootstrap.cluster.x-k8s.io,sideEffects=None
+
+var _ webhook.Validator = &EKSConfig{}
```
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/api/v1alpha3/eksconfig_webhook.go]: This has been removed :+1:

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 4 above)
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:192]: Oof, great point. This value needs to be the EKS cluster name from AWS' perspective - meaning it technically may not be the Cluster object's name unless we enforce that as a standard. 

Curious to get @richardcase's take on this as well.

**Comment by rudoi** [comment]
nothing like a 35-commit rebase to keep you on your toes!

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+			if rerr == nil {
+				rerr = err
+			}
+		}
+	}()
+
+	return r.joinWorker(ctx, scope)
+}
+
+func (r *EKSConfigReconciler) joinWorker(ctx context.Context, scope *EKSConfigScope) (ctrl.Result, error) {
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:157]:
```suggestion
func (r *EKSConfigReconciler) joinWorker(ctx context.Context, scope *EKSConfigScope) (ctrl.Result, error) {
		log := scope.Logger
```

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+	if scope.Config.Status.DataSecretName != nil {
+		secretKey := client.ObjectKey{Namespace: scope.Config.Namespace, Name: *scope.Config.Status.DataSecretName}
+		existingSecret := &corev1.Secret{}
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:160]:
```suggestion
		log = log.WithValues("data-secret-name", secretKey.Name)
		existingSecret := &corev1.Secret{}
```

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+
+		// No error here means the Secret exists and we have no
+		// reason to proceed.
+		err := r.Client.Get(ctx, secretKey, existingSecret)
+		switch {
+		case err == nil:
+			return ctrl.Result{}, nil
+		case !apierrors.IsNotFound(err):
+			return ctrl.Result{}, fmt.Errorf("unable to check for existing bootstrap secret: %w", err)
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:169]:
```suggestion
            log.Error(err, "unable to check for existing bootstrap secret")
			return ctrl.Result{}, err)
```

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 4 above)
Comment (richardcase) [reply bootstrap/eks/controllers/eksconfig_controller.go:192]: At the moment the cluster name is the same. I could suffix it with -eks to follow the conventions used for other resources?

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+
+	scope.Info("generating userdata")
+
+	// generate userdata
+	userDataScript, err := userdata.NewNode(&userdata.NodeInput{
+		ClusterName:      scope.Cluster.GetName(),
+		KubeletExtraArgs: scope.Config.Spec.KubeletExtraArgs,
+	})
+	if err != nil {
+		scope.Error(err, "Failed to create a worker join configuration")
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:196]:
```suggestion
		log.Error(err, "Failed to create a worker join configuration")
```

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+		conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretAvailableCondition, bootstrapv1.DataSecretGenerationFailedReason, clusterv1.ConditionSeverityWarning, "")
+		return ctrl.Result{}, err
+	}
+
+	// store userdata as secret
+	if err := r.storeBootstrapData(ctx, scope, userDataScript); err != nil {
+		scope.Error(err, "Failed to store bootstrap data")
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:203]:
```suggestion
		log.Error(err, "Failed to store bootstrap data")
```

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 21]
```diff
@@ -0,0 +1,357 @@
...
+		return ctrl.Result{}, nil
+	}
+
+	if !scope.Cluster.Status.ControlPlaneInitialized {
+		scope.Info("Control Plane has not yet been initialized")
+		conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretAvailableCondition, bootstrapv1.WaitingForControlPlaneInitializationReason, clusterv1.ConditionSeverityInfo, "")
+		return ctrl.Result{}, nil
+	}
+
+	scope.Info("generating userdata")
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:188]:
```suggestion
	log.Info("Generating userdata")
```

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 4 above)
Comment (richardcase) [reply bootstrap/eks/controllers/eksconfig_controller.go:192]: We could then add a func to the managed scope to get the underlying EKS cluster name which i will set in the status of the AWSManagedControlPlane. What do you think?

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (randomvariable) [reply bootstrap/eks/controllers/eksconfig_controller.go:126]: Think I'm in favour of cleaning this up as there's inconsistent use of the logger from either the reconciler or the scope (see below).

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 22]
```diff
@@ -0,0 +1,357 @@
...
+			"value": data,
+		},
+		Type: clusterv1.ClusterSecretType,
+	}
+
+	// as secret creation and scope.Config status patch are not atomic operations
+	// it is possible that secret creation happens but the config.Status patches are not applied
+	if err := r.Client.Create(ctx, secret); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "failed to create bootstrap data secret for EKSConfig %s/%s", scope.Config.Namespace, scope.Config.Name)
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:281]:
```suggestion
			return errors.Wrap(err, "failed to create bootstrap data secret for EKSConfig")
```
The caller will already append the keys via the logger values, so no need to add additional dynamic data.

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+		}
+		r.Log.Info("bootstrap data secret for EKSConfig already exists", "secret", secret.Name, "EKSConfig", scope.Config.Name)
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:283]: Would be better to pass in a logger from the caller that already has the keys.

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+	selectors := []client.ListOption{
+		client.InNamespace(c.Namespace),
+		client.MatchingLabels{
+			clusterv1.ClusterLabelName: c.Name,
+		},
+	}
+
+	machineList := &clusterv1.MachineList{}
+	if err := r.Client.List(context.Background(), machineList, selectors...); err != nil {
+		r.Log.Error(err, "failed to list Machines", "cluster", c.Name, "Namespace", c.Namespace)
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:344]:
```suggestion
		r.Log.Error(err, "failed to list Machines for Cluster", "name", c.Name, "namespace", c.Namespace)
```

File: test/e2e/go.mod [hunk 25]
```diff
@@ -12,13 +12,13 @@ require (
 	github.com/vmware-tanzu/sonobuoy v0.18.2
 	golang.org/x/sync v0.0.0-20200317015054-43a5402ce75a
 	golang.org/x/time v0.0.0-20200416051211-89c76fbcd5d1 // indirect
-	k8s.io/api v0.17.8
-	k8s.io/apimachinery v0.17.8
+	k8s.io/api v0.18.6
```
Comment (randomvariable) [inline test/e2e/go.mod:15]: Probably can ditch any changes to test/e2e?

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 26]
```diff
@@ -0,0 +1,357 @@
...
+
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=eksconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=eksconfigs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=awsclusters;awsmanagedclusters,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines;machinepools;clusters,verbs=get;list;watch;
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;delete;
+
+func (r *EKSConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rerr error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("namespace", req.NamespacedName.Namespace, "eksConfig", req.NamespacedName.Name)
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:72]:
```suggestion
	log := r.Log.WithValues("namespace", req.NamespacedName.Namespace, "name", req.NamespacedName.Name)
```

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 27]
```diff
@@ -0,0 +1,357 @@
...
+		// no error, requeue until we find an owner
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues(configOwner.GetKind(), configOwner.GetName())
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, configOwner.GetNamespace(), configOwner.ClusterName())
+	if err != nil {
+		if errors.Is(err, util.ErrNoCluster) {
+			log.Info(fmt.Sprintf("%s does not belong to a cluster yet, requeueing until it's part of a cluster", configOwner.GetKind()))
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:104]:
```suggestion
			log.Info("EKSConfig does not belong to a cluster yet, re-queuing until it's partof a cluster")
```

Already have keys.

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 28]
```diff
@@ -0,0 +1,357 @@
...
+			return ctrl.Result{}, nil
+		}
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster does not exist yet, requeueing until it is created")
```
Comment (randomvariable) [inline bootstrap/eks/controllers/eksconfig_controller.go:108]:
```suggestion
			log.Info("Cluster does not exist yet, re-queueing until it is created")
```

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 23 above)
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:283]: makes sense, i'm gunna batch commit all of your logging suggestions and then pass that through.

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:126]: sounds good, will do

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 4 above)
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go:192]: I think going with the Cluster's name is fine. Just wanted to confirm!

File: test/e2e/go.mod (hunk 25 above)
Comment (rudoi) [reply test/e2e/go.mod:15]: I think `make verify` will fail without this. I'll test...

Comment (rudoi) [reply test/e2e/go.mod:15]: you're right

File: bootstrap/eks/api/v1alpha3/eksconfig_webhook.go [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+// log is for logging in this package.
+var eksconfiglog = logf.Log.WithName("eksconfig-resource")
+
+func (r *EKSConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// (user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
```
[the remark this answers is not in the document]
Comment (rudoi) [reply bootstrap/eks/api/v1alpha3/eksconfig_webhook.go]: done

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 23 above)
Comment (rudoi) [reply bootstrap/eks/controllers/eksconfig_controller.go]: done

File: test/e2e/go.mod (hunk 25 above)
Comment (rudoi) [reply test/e2e/go.mod]: change removed - though `make modules` will put it back when run, FYI.

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 4 above)
Comment (randomvariable) [reply bootstrap/eks/controllers/eksconfig_controller.go]: Still worried about namespace clashes.

Comparing the acceptable characters of both:

AWS cluster name regex: ^[0-9A-Za-z][A-Za-z0-9\-_]*
Kubernetes: RFC1123
* contain no more than 253 characters
* contain only lowercase alphanumeric characters, '-' or '.'
* start with an alphanumeric character
* end with an alphanumeric character

That means someone could create a cluster with name "foo.bar" and this won't work.

Can I recommend that we use the following convention:

EKS cluster name: `<namespace>_<name>`, and with `.` also replaced with `_`?

Comment (randomvariable) [reply bootstrap/eks/controllers/eksconfig_controller.go]: There is also a 100 character limit for EKS, vs the 253 for k8s, so possibly also something similar to what we do for the ELB ( https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/services/elb/loadbalancer.go#L262 )

**Comment by randomvariable** [comment]
lgtm pending conclusion about what to do about translating cluster names to eks names.

File: bootstrap/eks/controllers/eksconfig_controller.go (hunk 4 above)
Comment (richardcase) [reply bootstrap/eks/controllers/eksconfig_controller.go]: Good point about namespace clashes. I'll change it like you mention about and hash if its over the 100 character limit.

**Comment by randomvariable** [comment]
/approve cancel

I'll do an approve after commit history is cleaned up.

File: bootstrap/eks/README.md [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+CABPE generates cloud-init configuration that CAPA will use as the [user-data](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html) for new EC2 instances. The generated configuration relies on the `bootstrap.sh` script that is present on all official AWS EKS AMIs. 
+
+The output script looks something like this (assuming an EKS cluster with the name `my-cluster`):
+
+```bash
+#!/bin/bash
+/etc/eks/bootstrap.sh my-cluster
+```
+
+CAPA will look up official the official EKS AMI for a particular Kubernetes version if the AMI is not overriden in the AWSMachine object. In order for this bootstrap provider to function, any user-specified AMIs will need to have `/etc/eks/bootstrap.sh` present (script source linked above).
```
Comment (richardcase) [inline bootstrap/eks/README.md:21]: Small nit `look up official the official EKS AMI` to `look up the official EKS AMI`

File: bootstrap/eks/api/v1alpha3/eksconfig_webhook.go [hunk 31]
```diff
@@ -0,0 +1,55 @@
+/*
```
Comment (richardcase) [inline bootstrap/eks/api/v1alpha3/eksconfig_webhook.go:1]: Could this be deleted as its not used?

File: bootstrap/eks/controllers/eksconfig_controller.go [hunk 32]
```diff
@@ -0,0 +1,344 @@
...
+// EKSConfigReconciler reconciles a EKSConfig object
+type EKSConfigReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=eksconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=eksconfigs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=awsclusters;awsmanagedclusters,verbs=get;list;watch;
```
Comment (richardcase) [inline bootstrap/eks/controllers/eksconfig_controller.go:58]: Could `awsmanagedclusters` be removed from the resources? I don't see it being used but i could well have missed it.

**Review by richardcase — APPROVED** [review]
A few very minor points.

**Comment by randomvariable** [comment]
```
--- a/test/e2e/go.mod
+++ b/test/e2e/go.mod
@@ -18,7 +18,7 @@ require (
 	k8s.io/utils v0.0.0-20200619165400-6e3d28b6ed19
 	sigs.k8s.io/cluster-api v0.3.7
 	sigs.k8s.io/cluster-api-provider-aws v0.5.3
-	sigs.k8s.io/controller-runtime v0.5.8
+	sigs.k8s.io/controller-runtime v0.5.9
 )
```

**Comment by randomvariable** [comment]
If we can get the fix and revert commits squashed, that'd be great too

**Comment by rudoi** [comment]
Fixing verify, missed a generate

**Review by richardcase — APPROVED** [review]
Great work 🎉
